src/compiler/checker.ts

```diff
@@ -35159,14 +35170,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             // since resolving such signature leads to resolving the potential outer signature, its arguments and thus the very same signature
             // it's possible that this inner resolution sets the resolvedSignature first.
             // In such a case we ignore the local result and reuse the correct one that was cached.
-            if (links.resolvedSignature !== resolvingSignature) {
+            // [cph 56013] This is buggy - links.resolvedSignature could be `undefined`. Unfortunately flow did not warn us.
+            // if (links.resolvedSignature !== resolvingSignature) {
+            //     result = links.resolvedSignature;
+            // }
+            if (links.resolvedSignature /* #56013 */ && links.resolvedSignature !== resolvingSignature) {
                 result = links.resolvedSignature;

```

Comment:
Fixed bug here - `links.resolvedSignature` is assumed to non-undefined, because flow assumes "no side effects".  But that only works if there are no side effects.  The bug that was fixed in this pull is based on eliminating a side effect between candidate iterations in `chooseOverload` - and that is done by resetting `getNodeLinks(node).resolvedSignature` to undefined.  

Perhaps the resetting in `checkExpression` could be changed to resetting to `resolvingSignature` instead of `undefined`?.   Then this extra check here would be unnecessary - but it would be worth adding a `Debug.assert(getNodeLinks(node).resolvedSignature)` immediately after the call to `resolveSignature()` to at least make clear that is a state invariant.  

However, is this logic:
```
if (links.resolvedSignature !== resolvingSignature) {
     result = links.resolvedSignature;
}
```
even necessary at all?  I will try removing it to test that question.